### PR TITLE
Any Base Exponentiation 

### PR DESCRIPTION
**Pull request by calebmkim**
This allows us to generate fixed point exponentiation, for any exponent and base. 

Sorry for the reaalllyy long PR, but a lot of it is tedious group building. 

At a high level, here is what I did: 

b^x = e^ln(b^x) = e ^ (x* ln(b)). 

As you can see, if we are able to calculate ln(b), then it is easy to see how to do this calculation. 

To calculate ln(b) we see that: 
ln(b) 
= ln(2^n * p), for some integer n and p between 1 and 2 
= n*ln(2) + ln(p). 
Therefore, if we can get a reliable value for ln(p)  we can get ln(b). I used a [Pade approximant](https://en.wikipedia.org/wiki/Pad%C3%A9_approximant) for ln(p). I don't really understand how it's calculated, but you can plug it into wolfram alpha [here](https://www.wolframalpha.com/input?i=+PadeApproximant%5Bln%28x%29%2C%7Bx%2C1.5%2C%7B2%2C2%7D%7D%5D+)) for a relatively easy, good, approximation of ln(p) when 1<=p<=2. 

I defined a component called `msb_calc` to calculate  n and p for a given value.

### CODE REVIEW

**Comment by sampsyo** [comment]
Wow; this looks incredibly cool. While the group-building does look tedious, the mathy approach looks sound. I say this is good to go.

This makes me thing of two much broader questions that are really not pertinent to the matter at hand:

* How can we make this kind of tedious group-building less tedious? Are there convenient Pythony things we can do to make it easier to write this type of Calyx-generating code?
* I wonder if there's room here for some kind of automated search for the best way to implement a mathematical expression. There are so many alternatives—different approximations for different primitives, different ways to group together associative operators, etc. That's not even to mention the tricky problem of different numerical representations! It seems like there's room here to do something cool to help find a good _strategy_ for implementing a given math expression.

**Comment by calebmkim** [comment]
Just wanted to check: do you think this is fine to merge or do you want to take more time to look at it? 

Reason I'm asking is that I've done some further work on this branch, and I got the `lrn` operation to work (bc of this new "any base" exponentiation), which makes GoogleNet possible, but I don't want to make this PR too big. 

As an aside: 
For your second question, I don't know how helpful it will be, but after working on these math operations the past few days, the basic strategy that has been used in Calyx so far seems to be the following: 

Separate the input we into an integer and (small) fixed point number (e.g., for e^x, simplify into e^(n+p), where n is an integer and p < 1. Or, for ln(x), simplify into ln(2^n * p) = n*ln(2) + ln(p), where n is an integer, and p is between one and two). 
The operation on the integer component should be easy, and then we find a good polynomial approximation for the fixed point number, which should be small so that the approximation works well (e.g., Taylor series for e^x, and Padé Approximant for ln,. I found out about Padé and other possible approximations by looking on Stack Exchange). 

Not sure how helpful this is, since it's more looking back on what strategies we ended up using and not necessarily what will work for future operations that we want.

File: calyx-py/calyx/gen_exp.py
Comment (cgyurgyik) [inline calyx-py/calyx/gen_exp.py:994]: Doesn't `any_base_program` also include `e`? Why did you make this decision?

File: calyx-py/calyx/gen_exp.py
Comment (cgyurgyik) [inline calyx-py/calyx/gen_exp.py:995]: Is this all essentially thrown away if the `base_is_e` flag is false? Maybe you want to place this under the if/else as well, to avoid that (perhaps in a new function).

**Comment by cgyurgyik** [comment]
> I wonder if there's room here for some kind of automated search for the best way to implement a mathematical expression. There are so many alternatives—different approximations for different primitives, different ways to group together associative operators, etc. That's not even to mention the tricky problem of different numerical representations! It seems like there's room here to do something cool to help find a good strategy for implementing a given math expression.

That'd be extremely useful, a single place to look at all the (currently known) ways to implement a mathematical function, incorporating error, limiting behavior in software, hardware requirements, etc.

**Review by cgyurgyik — COMMENTED** [review]
Cool work :)

File: calyx-py/calyx/gen_exp.py
Comment (calebmkim) [reply calyx-py/calyx/gen_exp.py:995]: Yeah, it' probably cleaner to do it that way. I should update that on this PR before we merge it.

File: calyx-py/calyx/gen_exp.py
Comment (calebmkim) [reply calyx-py/calyx/gen_exp.py:994]: The way we calculate b^x (for any base b) is by doing e^ln(b^x). So, we need to generate a lot of extra components to do the ln(b^x). But if the base is just `e` then we don't need all those extra components.

Comment (cgyurgyik) [reply calyx-py/calyx/gen_exp.py:994]: Got it. I would suggest adding a comment to clarify this. Alternatively just use 'base_is_e' and 'not base_is_e'. Since any base seems like a misnomer. Up to you :)

Comment (cgyurgyik) [reply calyx-py/calyx/gen_exp.py:994]: On second thought, can we infer the base is 'e' when no base is provided, instead of adding an additional flag?

**Review by sampsyo — APPROVED** [review]
Yes, I do think this is good to go! No need to hold off once you're satisfied, IMO—there will always be more to do on math functions.

File: calyx-py/calyx/gen_exp.py
Comment (calebmkim) [reply calyx-py/calyx/gen_exp.py:994]: I think it'll be easier to clear this up on a subsequent PR (I'll make it later today).

**Pull request merged**
